src/StringTools/OpportunisticIntern.cs

```diff
@@ -0,0 +1,295 @@
+                    result.AppendLine("Elimination assumes that strings provided were unique objects.");
+                    result.AppendLine("|---------------------------------------------------------------------------------|");
+
+                    IEnumerable<string> topMissingHardcodedString =
+                        _missedHardcodedStrings
+                        .OrderByDescending(kv => kv.Value * kv.Key.Length)
+                        .Take(15)
+                        .Where(kv => kv.Value > 1)
+                        .Select(kv => string.Format(CultureInfo.InvariantCulture, "({1} instances x each {2} chars)\n{0}", kv.Key, kv.Value, kv.Key.Length));
+
+                    result.AppendLine(string.Format("##########Top Missing Hardcoded Strings:  \n{0} ", string.Join("\n==============\n", topMissingHardcodedString.ToArray())));
+                    result.AppendLine();
+
+                    WeakStringCache.DebugInfo debugInfo = _weakStringCache.GetDebugInfo();
+                    result.AppendLine("WeakStringCache statistics:");
+                    result.AppendLine(string.Format("String count live/collected/total = {0}/{1}/{2}", debugInfo.LiveStringCount, debugInfo.CollectedStringCount, debugInfo.LiveStringCount + debugInfo.CollectedStringCount));
+                }
+                else
+                {
+                    result.Append(title);
+                    result.AppendLine(" - EnableStatisticsGathering() has not been called");
+                }
+
+                return result.ToString();
+            }
+
+            /// <summary>
+            /// Try to intern the string.
+            /// The return value indicates the how the string was interned (if at all).
+            /// </summary>
+            private InternResult TryIntern(ref InternableString candidate, out string interned)
+            {
+                // First, try the interning callbacks.
+                if (Strings.CallStringInterningCallbacks(ref candidate, out interned))
+                {
+                    // Either matched a hardcoded string or is explicitly not to be interned.
+                    if (interned != null)
+                    {
+                        return InternResult.MatchedHardcodedString;
+                    }
+                    interned = candidate.ExpensiveConvertToString();
+                    return InternResult.RejectedFromInterning;
+                }
+
+                interned = _weakStringCache.GetOrCreateEntry(ref candidate, out bool cacheHit);
+                return cacheHit ? InternResult.FoundInWeakStringCache : InternResult.AddedToWeakStringCache;
+            }
+
+            /// <summary>
+            /// Version of Intern that gathers statistics
+            /// </summary>
+            private string InternWithStatistics(ref InternableString candidate)
+            {
+                lock (_missedHardcodedStrings)
```

Comment:
Does the lock need to be over this whole section or just the if statement below?

Reply:
This code has moved but I believe that your question still applies. In the current version of the code the lock protects several fields: `_regularInternHits`, `_regularInternMisses`, `_internCallCountsByString`, `_internEliminatedStrings`, `_internEliminatedChars`. Integer fields could be accessed lock-free with interlocked operations but given that we're also mutating a dictionary, it's easier to put everything under one lock. The code is used only for gathering statistics for debugging so it's not super perf sensitive.